[JavaScriptCore] Build with InstallAPI

emw-apple opened the pull request:
#### a9c99ebb0bfd3a03a7f532d39f9a18f984c9a8b0
<pre>
[JavaScriptCore] Build with InstallAPI
<a href="https://bugs.webkit.org/show_bug.cgi?id=254784">https://bugs.webkit.org/show_bug.cgi?id=254784</a>
rdar://108843961

Reviewed by Alexey Proskuryakov.

InstallAPI is a build phase in Xcode that uses tapi(1) to create API
stubs for framework targets by parsing their headers. These .tbd files
can then be linked against by clients of the framework, such that
clients are able to build and link before the framework itself has
finished building.

We already use InstallAPI in iOS WebKit/WebKitLegacy. Enable it in
JavaScriptCore.

Because JavaScriptCore exports the WTF and bmalloc static libraries,
TAPI needs to read headers from those projects in addition to its own. Add
a script to bmalloc and WTF which generates header file lists based on
all-product-headers.yaml, which is metadata created by XCBuild that
contains the public/private headers of a target. Changes to the headers
being installed will rerun both the file list generator and InstallAPI.

In production-style builds, all-products-headers.yaml is not available
during the installhdrs phase, so the script produces its file list by
scanning the headers directory being installed. This doesn&apos;t carry the
staleness risk of scanning directories in normal builds, because
install-style builds are always clean.

Also in production-style builds, the SDKROOT path, which is referenced
in generated filelists, may change when building different projects. We
handle this by writing &quot;&lt;SDKROOT&gt;&quot; in path names when the filelists are
generated, and then replacing that token with the actual SDKROOT right
before JavaScriptCore invokes tapi.

* Source/JavaScriptCore/API/ExtraSymbolsForTAPI.h: Added, lists a few
  symbols that really shouldn&apos;t be declared in any headers.
* Source/JavaScriptCore/Configurations/Base.xcconfig: JavaScriptCore
  includes private OS headers from inside System.framework, but was
  adding this path using OTHER_CPLUSPLUSFLAGS. Move it to
  SYSTEM_HEADER_SEARCH_PATHS so that TAPI gets the search path, too.
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig: Enable
  TAPI and add flags for dialect settings and extra filelists.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj: Add
  reference to ExtraSymbolsForTAPI.h. Remove LLIntAssembly.h&apos;s target
  membership, because it is not generated in installapi style builds and
  is not needed by tapi.
* Source/WTF/Configurations/WTF.xcconfig: Enable
  INSTALLHDRS_SCRIPT_PHASE to invoke generate-tapi-filelist.py during
  installhdrs. Remove INSTALLHDRS_COPY_PHASE as headers were moved out
  of Copy Files phases last year.
* Source/WTF/Scripts/generate-tapi-filelist.py: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj: Add build phase to generate
  a TAPI filelist using the new script.
* Source/bmalloc/Configurations/bmalloc.xcconfig: Turn on
  INSTALLHDRS_SCRIPT_PHASE to invoke generate-tapi-filelist.py during
  installhdrs.
* Source/bmalloc/Configurations/libpas.xcconfig: Build with
  -DPAS_BMALLOC_HIDDEN, needed to fix symbol visibility when building
  with ThinLTO (rdar://86242029).
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj: Ditto pbxproj
  changes.
* Tools/Scripts/webkitpy/style/checkers/cpp.py: Allow bmalloc and WTF
  export macros in Source/JavaScriptCore/API/ExtraSymbolsForTAPI.h.

Fix up some generated headers&apos; target membership to make tapi succeed.
These headers are still tracked in the project, but are either removed
from the target or no longer installed.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project/pbxproj:
  - Remove LLIntAssembly.h&apos;s target membership. It is not generated
    during the installapi phase and is not needed by tapi.
  - Remove target membership from AirOpcodeGenerated.h and
    AirOpcodeUtils.h. They are missing some includes when they are
    processed by tapi, and they do not contain any API or SPI.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj: Change
  thingy_heap.h&apos;s visibility to Project, because one of the headers it
  includes, thingy_heap_prefix.h, is project-only. This never broke a
  build because none of bmalloc&apos;s clients actually use this header.

Fix a number of symbol visibility discrepencies to make tapi succeed.

* Source/JavaScriptCore/assembler/JITOperationList.h: Declare
  populatePointersInEmbedder inside an ENABLE(JIT_OPERATION_VALIDATION)
  block, matching its implementation.
* Source/JavaScriptCore/dfg/DFGDoesGCCheck.h: Declare verifyCanGC inside
  a ENABLE(DFG_DOES_GC_VALIDATION) block, matching its implementation.
* Source/JavaScriptCore/runtime/VM.h: Move ENABLE_DFG_DOES_GC_VALIDATION
  definition to PlatformEnable.h so that it can be used in
  DFGDoesGCCheck.h.
* Source/WTF/wtf/Compiler.h: Silence error macro about
  -DRELEASE_WITHOUT_OPTIMIZATIONS.
* Source/WTF/wtf/PlatformEnable.h: Add ENABLE_DFG_DOES_GC_VALIDATION
  from &lt;JavaScriptCore/VM.h&gt;, see above.
* Source/WTF/wtf/PtrTag.h: Declare tag debugging symbols inside an
  ENABLE(PTRTAG_DEBUGGING) block, matching their implementation.
* Source/bmalloc/libpas/src/libpas/pas_report_crash.h: Correct a manual
  visibility attribute to be PAS_API, so that the declaration follows
  PAS_BMALLOC_HIDDEN. Also, fix style errors on this line.

Additionally, add missing includes and visibility decrepencies for
non-JIT configurations of JavaScriptCore, namely the watchOS build.

* Source/JavaScriptCore/b3/air/opcode_generator.rb: Surround generated
  AirOpcode.h header with ENABLE(B3_JIT).
* Source/JavaScriptCore/bytecode/FullBytecodeLiveness.h: Add include
  needed by FuleBytecodeLiveness::virtualRegisterIsLive inline
  definition.
* Source/JavaScriptCore/dfg/DFGArgumentPosition.h: Surround with
  ENABLE(DFG_JIT).
* Source/JavaScriptCore/dfg/DFGForAllKills.h: Surround with
  ENABLE(DFG_JIT).
* Source/JavaScriptCore/dfg/DFGGraph.cpp: Surround with ENABLE(DFG_JIT).
* Source/JavaScriptCore/dfg/DFGMinifiedIDInlines.h: Surround with
  ENABLE(DFG_JIT).
* Source/JavaScriptCore/ftl/FTLLazySlowPathCall.h: Surround with
  ENABLE(FTL_JIT).
* Source/JavaScriptCore/runtime/JSArrayBufferPrototypeInlines.h: Add
  needed include for parsing JSArrayBuffer, which is only implicitly
  included when building with JIT.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj: LineMetadata.h and
  HeapConstants.h declare API when libpas is disabled. Give the
  &quot;private&quot; membership so that tapi sees them.

Canonical link: <a href="https://commits.webkit.org/263845@main">https://commits.webkit.org/263845@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c92dc78583341eac97dff5b95d2910a5c6c520f8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5828 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5996 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6185 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7388 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6215 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5829 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6219 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5963 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/8127 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5934 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5984 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5282 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7445 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/5936 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3461 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5257 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13219 "2 flakes 188 failures") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/4873 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5328 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5336 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7565 "run-api-tests-without-change (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/5420 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5781 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4747 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/5945 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5222 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1444 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1392 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9343 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/6110 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5583 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1564 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/JavaScriptCore/Configurations/Base.xcconfig:
```diff
@@ -95,7 +95,7 @@ WARNING_CFLAGS = $(inherited) -Wcast-qual -Wchar-subscripts -Wextra-tokens -Wini
 HEADER_SEARCH_PATHS = . "$(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH)" $(HEADER_SEARCH_PATHS);
 LIBRARY_SEARCH_PATHS = $(SDK_DIR)$(WK_LIBRARY_INSTALL_PATH) $(inherited);
-SYSTEM_HEADER_SEARCH_PATHS = $(SDK_DIR)$(WK_ALTERNATE_WEBKIT_SDK_PATH)$(WK_LIBRARY_HEADERS_FOLDER_PATH) $(inherited);
+SYSTEM_HEADER_SEARCH_PATHS = $(SDK_DIR)$(WK_ALTERNATE_WEBKIT_SDK_PATH)$(WK_LIBRARY_HEADERS_FOLDER_PATH) -isystem $(SDKROOT)/System/Library/Frameworks/System.framework/PrivateHeaders $(inherited);
```
`-isystem` seems like a copy/paste mistake.

emw-apple commented on Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
```diff
@@ -11927,6 +11924,28 @@
+			);
+			runOnlyForDeploymentPostprocessing = 1;
+			shellPath = /bin/sh;
+			shellScript = "for ((i=0; i<${SCRIPT_INPUT_FILE_COUNT}; i++)); do\n    eval input=\"\\${SCRIPT_INPUT_FILE_$i}\"\n    eval output=\"\\${SCRIPT_OUTPUT_FILE_$i}\"\n    sed -e \"s;<SDKROOT>;${SDKROOT};\" $input > $output\ndone\n";
```
Calling attention to one more change I discovered while doing internal testing.

> Also in production-style builds, the SDKROOT path, which is referenced
> in generated filelists, may change when building different projects. We
> handle this by writing "\<SDKROOT\>" in path names when the filelists are
> generated, and then replacing that token with the actual SDKROOT right
> before JavaScriptCore invokes tapi.

Here's the script expanded:
```sh
for ((i=0; i<${SCRIPT_INPUT_FILE_COUNT}; i++)); do
    eval input="\${SCRIPT_INPUT_FILE_$i}"
    eval output="\${SCRIPT_OUTPUT_FILE_$i}"
    sed -e "s;<SDKROOT>;${SDKROOT};" $input > $output
done
```

emw-apple committed:
[JavaScriptCore] Build with InstallAPI
https://bugs.webkit.org/show_bug.cgi?id=254784
rdar://108843961

Reviewed by Alexey Proskuryakov.

InstallAPI is a build phase in Xcode that uses tapi(1) to create API
stubs for framework targets by parsing their headers. These .tbd files
can then be linked against by clients of the framework, such that
clients are able to build and link before the framework itself has
finished building.

We already use InstallAPI in iOS WebKit/WebKitLegacy. Enable it in
JavaScriptCore.

Because JavaScriptCore exports the WTF and bmalloc static libraries,
TAPI needs to read headers from those projects in addition to its own. Add
a script to bmalloc and WTF which generates header file lists based on
all-product-headers.yaml, which is metadata created by XCBuild that
contains the public/private headers of a target. Changes to the headers
being installed will rerun both the file list generator and InstallAPI.

In production-style builds, all-products-headers.yaml is not available
during the installhdrs phase, so the script produces its file list by
scanning the headers directory being installed. This doesn't carry the
staleness risk of scanning directories in normal builds, because
install-style builds are always clean.

Also in production-style builds, the SDKROOT path, which is referenced
in generated filelists, may change when building different projects. We
handle this by writing "<SDKROOT>" in path names when the filelists are
generated, and then replacing that token with the actual SDKROOT right
before JavaScriptCore invokes tapi.

* Source/JavaScriptCore/API/ExtraSymbolsForTAPI.h: Added, lists a few
  symbols that really shouldn't be declared in any headers.
* Source/JavaScriptCore/Configurations/Base.xcconfig: JavaScriptCore
  includes private OS headers from inside System.framework, but was
  adding this path using OTHER_CPLUSPLUSFLAGS. Move it to
  SYSTEM_HEADER_SEARCH_PATHS so that TAPI gets the search path, too.
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig: Enable
  TAPI and add flags for dialect settings and extra filelists.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj: Add
  reference to ExtraSymbolsForTAPI.h. Remove LLIntAssembly.h's target
  membership, because it is not generated in installapi style builds and
  is not needed by tapi.
* Source/WTF/Configurations/WTF.xcconfig: Enable
  INSTALLHDRS_SCRIPT_PHASE to invoke generate-tapi-filelist.py during
  installhdrs. Remove INSTALLHDRS_COPY_PHASE as headers were moved out
  of Copy Files phases last year.
* Source/WTF/Scripts/generate-tapi-filelist.py: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj: Add build phase to generate
  a TAPI filelist using the new script.
* Source/bmalloc/Configurations/bmalloc.xcconfig: Turn on
  INSTALLHDRS_SCRIPT_PHASE to invoke generate-tapi-filelist.py during
  installhdrs.
* Source/bmalloc/Configurations/libpas.xcconfig: Build with
  -DPAS_BMALLOC_HIDDEN, needed to fix symbol visibility when building
  with ThinLTO (rdar://86242029).
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj: Ditto pbxproj
  changes.
* Tools/Scripts/webkitpy/style/checkers/cpp.py: Allow bmalloc and WTF
  export macros in Source/JavaScriptCore/API/ExtraSymbolsForTAPI.h.

Fix up some generated headers' target membership to make tapi succeed.
These headers are still tracked in the project, but are either removed
from the target or no longer installed.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project/pbxproj:
  - Remove LLIntAssembly.h's target membership. It is not generated
    during the installapi phase and is not needed by tapi.
  - Remove target membership from AirOpcodeGenerated.h and
    AirOpcodeUtils.h. They are missing some includes when they are
    processed by tapi, and they do not contain any API or SPI.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj: Change
  thingy_heap.h's visibility to Project, because one of the headers it
  includes, thingy_heap_prefix.h, is project-only. This never broke a
  build because none of bmalloc's clients actually use this header.

Fix a number of symbol visibility discrepencies to make tapi succeed.

* Source/JavaScriptCore/assembler/JITOperationList.h: Declare
  populatePointersInEmbedder inside an ENABLE(JIT_OPERATION_VALIDATION)
  block, matching its implementation.
* Source/JavaScriptCore/dfg/DFGDoesGCCheck.h: Declare verifyCanGC inside
  a ENABLE(DFG_DOES_GC_VALIDATION) block, matching its implementation.
* Source/JavaScriptCore/runtime/VM.h: Move ENABLE_DFG_DOES_GC_VALIDATION
  definition to PlatformEnable.h so that it can be used in
  DFGDoesGCCheck.h.
* Source/WTF/wtf/Compiler.h: Silence error macro about
  -DRELEASE_WITHOUT_OPTIMIZATIONS.
* Source/WTF/wtf/PlatformEnable.h: Add ENABLE_DFG_DOES_GC_VALIDATION
  from <JavaScriptCore/VM.h>, see above.
* Source/WTF/wtf/PtrTag.h: Declare tag debugging symbols inside an
  ENABLE(PTRTAG_DEBUGGING) block, matching their implementation.
* Source/bmalloc/libpas/src/libpas/pas_report_crash.h: Correct a manual
  visibility attribute to be PAS_API, so that the declaration follows
  PAS_BMALLOC_HIDDEN. Also, fix style errors on this line.

Additionally, add missing includes and visibility decrepencies for
non-JIT configurations of JavaScriptCore, namely the watchOS build.

* Source/JavaScriptCore/b3/air/opcode_generator.rb: Surround generated
  AirOpcode.h header with ENABLE(B3_JIT).
* Source/JavaScriptCore/bytecode/FullBytecodeLiveness.h: Add include
  needed by FuleBytecodeLiveness::virtualRegisterIsLive inline
  definition.
* Source/JavaScriptCore/dfg/DFGArgumentPosition.h: Surround with
  ENABLE(DFG_JIT).
* Source/JavaScriptCore/dfg/DFGForAllKills.h: Surround with
  ENABLE(DFG_JIT).
* Source/JavaScriptCore/dfg/DFGGraph.cpp: Surround with ENABLE(DFG_JIT).
* Source/JavaScriptCore/dfg/DFGMinifiedIDInlines.h: Surround with
  ENABLE(DFG_JIT).
* Source/JavaScriptCore/ftl/FTLLazySlowPathCall.h: Surround with
  ENABLE(FTL_JIT).
* Source/JavaScriptCore/runtime/JSArrayBufferPrototypeInlines.h: Add
  needed include for parsing JSArrayBuffer, which is only implicitly
  included when building with JIT.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj: LineMetadata.h and
  HeapConstants.h declare API when libpas is disabled. Give the
  "private" membership so that tapi sees them.

Canonical link: https://commits.webkit.org/263845@main
